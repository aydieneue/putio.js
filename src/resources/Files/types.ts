export type FileType =
  | 'FOLDER'
  | 'FILE'
  | 'AUDIO'
  | 'VIDEO'
  | 'IMAGE'
  | 'ARCHIVE'
  | 'PDF'
  | 'TEXT'
  | 'SWF'
