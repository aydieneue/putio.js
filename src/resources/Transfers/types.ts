import { ISODateString } from '../../utils/types'

type TransferType = 'URL' | 'TORRENT' | 'PLAYLIST' | 'LIVE_STREAM'

type TransferStatus =
  | 'WAITING'
  | 'IN_QUEUE'
  | 'DOWNLOADING'
  | 'COMPLETING'
  | 'STOPPING'
  | 'SEEDING'
  | 'COMPLETED'
  | 'ERROR'
  | 'PREPARING_SEED'

type BaseTransfer = {
  availability: null | number
  created_at: ISODateString
  current_ratio: null | number
  down_speed: number
  downloaded: number
  error_message: null | string
  estimated_time: null | number
  file_id: null | number
  finished_at: null | ISODateString
  id: number
  is_private: boolean
  name: string
  save_parent_id: number
  seconds_seeding: null | number
  size: number
  source: string
  started_at: null | ISODateString
  status: TransferStatus
  type: TransferType
  subscription_id: null | number
  uploaded: number
  up_speed: number

  // not documented in swagger
  callback_url: null | string
  client_ip: null | string
  completion_percent: number // used?
  created_torrent: boolean // used?
  download_id: number
  hash: null | string // used?
  links: string[] // used?
  peers_connected: number
  peers_getting_from_us: number
  peers_sending_to_us: number
  percent_done: number
  simulated: boolean
  status_message: string
  torrent_link: null | string
  tracker: null | string
  tracker_message: null | string

  // documented in swagger but doesn't come in response, delete from api docs?
  peers: number
}

type FinishedTransfer = BaseTransfer & {
  finished_at: ISODateString
  file_id: number
}

/* by status */
type CompletedTransfer = FinishedTransfer & {
  status: 'COMPLETED'
}

type FailedTransfer = BaseTransfer & {
  status: 'ERROR'
  error_message: string
}

/* by status, but scoped to type */
type PreparingSeedingTorrentTransfer = FinishedTransfer & {
  type: 'TORRENT'
  status: 'PREPARING_SEED'
}

type SeedingTorrentTransfer = FinishedTransfer & {
  type: 'TORRENT'
  status: 'SEEDING'
  seconds_seeding: number
  current_ratio: number
}

type CompletedTorrentTransfer = SeedingTorrentTransfer & {
  status: 'COMPLETED'
}

/* by type */
export type TorrentTransfer = (
  | BaseTransfer
  | FailedTransfer
  | SeedingTorrentTransfer
  | PreparingSeedingTorrentTransfer
  | CompletedTorrentTransfer
) & {
  type: 'TORRENT'
  torrent_link: string
}

export type URLTransfer = (
  | BaseTransfer
  | FailedTransfer
  | CompletedTransfer
) & {
  type: 'URL'
}

export type PlaylistTransfer = (
  | BaseTransfer
  | FailedTransfer
  | CompletedTransfer
) & {
  type: 'PLAYLIST'

  // not documented in swagger. not present in transfers/list response, but comes within socket updates
  downloaded_items: number
  total_items: number

  // why present in this one?
  recorded_seconds: number
}

export type LiveStreamTransfer = (
  | BaseTransfer
  | FailedTransfer
  | CompletedTransfer
) & {
  type: 'LIVE_STREAM'

  // not documented in swagger
  recorded_seconds: number
}

export type Transfer =
  | TorrentTransfer
  | URLTransfer
  | PlaylistTransfer
  | LiveStreamTransfer

export interface ITransfersResponse {
  transfers: Transfer[]
  total: number
  cursor: string
}
