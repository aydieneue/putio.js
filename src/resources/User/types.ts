export interface IUserProperties extends Record<string, any> {
  username: string
  account_active: boolean
  files_will_be_deleted_at: string | null
  is_admin: boolean
  disk: {
    avail: number
    used: number
    size: number
  }
}

export interface IUserSettings extends Record<string, any> {
  show_optimistic_usage: boolean
  disable_subtitles: boolean
}

export interface IUser extends IUserProperties {
  settings: IUserSettings
}

export interface IUserInfoResponse {
  info: IUser
}

export interface IUserSettingsResponse {
  settings: IUserSettings
}
