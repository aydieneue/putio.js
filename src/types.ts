import { AxiosError, AxiosResponse } from "axios";
import PutioAPIClient from ".";

export interface IPutioAPIClientOptions {
  clientID?: number;
  baseURL?: string;
}

export interface IPutioAPIClientResponse extends AxiosResponse {
  body?: any; // Remove when it's irrelevant.
}

interface IPutioAPIClientErrorData {
  error_id?: string;
  error_uri?: string;
  error_type: string;
  error_message: string;
  status_code: number;
}
export interface IPutioAPIClientError extends AxiosError {
  data?: IPutioAPIClientErrorData;
}

export interface IPutioAPIClientMiddleware {
  onFulfilled: (response: IPutioAPIClientResponse) => IPutioAPIClientResponse;
  onRejected: (error: IPutioAPIClientError) => Promise<IPutioAPIClientError>;
}

export type IPutioAPIClientMiddlewareFactory = (
  client: PutioAPIClient
) => IPutioAPIClientMiddleware;

export enum PutioAPIClientEventTypes {
  ERROR = "ERROR",
  CLIENT_IP_CHANGED = "CLIENT_IP_CHANGED"
}
