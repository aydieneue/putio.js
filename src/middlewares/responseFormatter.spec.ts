import {
  MockPutioAPIClient,
  mockPutioAPIClientError,
  mockPutioAPIClientResponse,
} from '../test-utils/mocks'
import { IPutioAPIClientError } from '../types'
import create from './responseFormatter'

describe('middlewares/responseFormatter', () => {
  const API = new MockPutioAPIClient()
  const responseFormatter = create(API)

  describe('successful responses', () => {
    it('transforms as expected', () => {
      expect(
        responseFormatter.onFulfilled(mockPutioAPIClientResponse),
      ).toMatchSnapshot()
    })
  })

  describe('failed responses', () => {
    it('sets error.data property correctly when the request failed with Put.io API signature', () => {
      const error = {
        ...mockPutioAPIClientError,
        response: {
          config: {},
          data: {
            error_type: 'API_ERROR',
            error_message: 'Putio API Error',
          },
          headers: {},
          status: 400,
          statusText: 'Error!',
        },
      }

      responseFormatter
        .onRejected(error)
        .catch(e => expect(e).toMatchSnapshot())
    })

    it('sets error.data property correctly when the request failed with HTTP response but without Put.io API signature', () => {
      const error: IPutioAPIClientError = {
        ...mockPutioAPIClientError,
        response: {
          config: {},
          headers: {},
          data: 'Bad Gateway',
          status: 502,
          statusText: 'Bad Gateway',
        },
      }

      responseFormatter
        .onRejected(error)
        .catch(e => expect(e).toMatchSnapshot())
    })

    it('sets error.data property correctly when the request failed without Put.io API signature', () => {
      responseFormatter
        .onRejected(mockPutioAPIClientError)
        .catch(e => expect(e).toMatchSnapshot())
    })
  })
})
