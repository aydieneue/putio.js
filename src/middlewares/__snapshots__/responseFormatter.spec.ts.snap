// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/responseFormatter failed responses sets error.data property correctly when the request failed with HTTP response but without Put.io API signature 1`] = `
Object {
  "config": Object {},
  "data": Object {
    "error_message": "Mock Error!",
    "error_type": "ERROR",
    "status_code": 502,
  },
  "isAxiosError": true,
  "message": "Mock Error!",
  "name": "MockError",
  "response": Object {
    "config": Object {},
    "data": "Bad Gateway",
    "headers": Object {},
    "status": 502,
    "statusText": "Bad Gateway",
  },
}
`;

exports[`middlewares/responseFormatter failed responses sets error.data property correctly when the request failed with Put.io API signature 1`] = `
Object {
  "config": Object {},
  "data": Object {
    "error_message": "Putio API Error",
    "error_type": "API_ERROR",
    "status_code": 400,
  },
  "isAxiosError": true,
  "message": "Mock Error!",
  "name": "MockError",
  "response": Object {
    "config": Object {},
    "data": Object {
      "error_message": "Putio API Error",
      "error_type": "API_ERROR",
    },
    "headers": Object {},
    "status": 400,
    "statusText": "Error!",
  },
}
`;

exports[`middlewares/responseFormatter failed responses sets error.data property correctly when the request failed without Put.io API signature 1`] = `
Object {
  "config": Object {},
  "data": Object {
    "error_message": "Mock Error!",
    "error_type": "ERROR",
    "status_code": 0,
  },
  "isAxiosError": true,
  "message": "Mock Error!",
  "name": "MockError",
}
`;

exports[`middlewares/responseFormatter succesfull responses transforms as expected 1`] = `
Object {
  "body": Object {
    "hello": "world",
  },
  "config": Object {},
  "data": Object {
    "hello": "world",
  },
  "headers": Object {},
  "status": 200,
  "statusText": "ok",
}
`;
